File: tests/mock/src/contract/canonical.ts
import type { SigningRequest } from './types.ts'
import { sha256 } from '@noble/hashes/sha2.js'
import { InvalidEnvelopeError } from './errors.ts'

// Byte-exact TypeScript port of internal/contract/canonical.go
// (docs/spec/envelope-canonical.md §2). The committed testdata/golden.json,
// produced by the real Go package, is the authoritative cross-language anchor
// asserted in test/golden.test.ts — this file MUST reproduce those bytes.

/**
 * DOMAIN = ASCII("TSS-ENVELOPE-CANONICAL-v1") ‖ 0x00 (§2.2). The trailing NUL
 * terminates the constant and versions the scheme (orthogonal to
 * SigningRequest.version).
 */
export const CANONICAL_DOMAIN: Uint8Array = (() => {
  const tag = new TextEncoder().encode('TSS-ENVELOPE-CANONICAL-v1')
  const out = new Uint8Array(tag.length + 1)
  out.set(tag, 0)
  out[tag.length] = 0x00
  return out
})()

const FIXED_HASH_LEN = 32
const MAX_LEN_PREFIX = 0xFFFFFFFF

function nfcBytes(s: string): Uint8Array {
  // §2.3 rule 2: Unicode NFC then UTF-8, so visually identical strings from
  // different input methods canonicalize identically across parties.
  return new TextEncoder().encode(s.normalize('NFC'))
}

/**
 * Parse an RFC 4122 UUID string into its 16 raw bytes (§2.3). Accepts the
 * hyphenated 8-4-4-4-12 form and the 32-hex unhyphenated form, hex
 * case-insensitive; anything else is an invalid envelope.
 */
export function uuidBytes(s: string): Uint8Array {
  let hexStr: string
  if (s.length === 36) {
    if (s[8] !== '-' || s[13] !== '-' || s[18] !== '-' || s[23] !== '-')
      throw new InvalidEnvelopeError('malformed requestId UUID')
    hexStr = s.slice(0, 8) + s.slice(9, 13) + s.slice(14, 18) + s.slice(19, 23) + s.slice(24)
  }
  else if (s.length === 32) {
    hexStr = s
  }
  else {
    throw new InvalidEnvelopeError(`requestId is not a UUID (len ${s.length})`)
  }
  if (!/^[0-9a-f]{32}$/i.test(hexStr))
    throw new InvalidEnvelopeError('requestId not hex')
  const out = new Uint8Array(16)
  for (let i = 0; i < 16; i++)
    out[i] = Number.parseInt(hexStr.slice(i * 2, i * 2 + 2), 16)
  return out
}

class ByteBuf {
  private chunks: Uint8Array[] = []

  write(b: Uint8Array): void {
    this.chunks.push(b)
  }

  writeUint64(v: number | bigint): void {
    const buf = new Uint8Array(8)
    new DataView(buf.buffer).setBigUint64(0, BigInt(v), false)
    this.chunks.push(buf)
  }

  // writeInt64: Go writes uint64(v) two's-complement big-endian; setBigInt64
  // produces the identical bytes for the int64 range.
  writeInt64(v: number | bigint): void {
    const buf = new Uint8Array(8)
    new DataView(buf.buffer).setBigInt64(0, BigInt(v), false)
    this.chunks.push(buf)
  }

  writeLenPrefixed(field: string, b: Uint8Array): void {
    if (b.length > MAX_LEN_PREFIX)
      throw new InvalidEnvelopeError(`${field} exceeds uint32 length`)
    const lp = new Uint8Array(4)
    new DataView(lp.buffer).setUint32(0, b.length, false)
    this.chunks.push(lp, b)
  }

  bytes(): Uint8Array {
    let n = 0
    for (const c of this.chunks) n += c.length
    const out = new Uint8Array(n)
    let off = 0
    for (const c of this.chunks) {
      out.set(c, off)
      off += c.length
    }
    return out
  }
}

/**
 * Build the canonical preimage P of an envelope (§2.1):
 *
 *   P = DOMAIN ‖ F(version) ‖ F(requestId) ‖ F(groupId) ‖ F(chain)
 *         ‖ F(unsignedTx) ‖ F(digest32) ‖ F(proposer)
 *         ‖ F(createdAt) ‖ F(expiry) ‖ F(metaHash)
 *
 * businessInfo never enters P (carried via metaHash); proposerSig never enters
 * P (it signs P). Bytes are derived from logical values only.
 */
export function canonicalBytes(req: SigningRequest): Uint8Array {
  const buf = new ByteBuf()
  buf.write(CANONICAL_DOMAIN)
  buf.writeUint64(req.version)
  buf.write(uuidBytes(req.requestId))
  buf.writeLenPrefixed('groupId', nfcBytes(req.groupId))
  buf.writeLenPrefixed('chain', nfcBytes(req.chain))
  buf.writeLenPrefixed('unsignedTx', req.unsignedTx)

  if (req.digest32.length !== FIXED_HASH_LEN)
    throw new InvalidEnvelopeError(`digest32 is ${req.digest32.length} bytes, want ${FIXED_HASH_LEN}`)
  buf.write(req.digest32)

  buf.writeLenPrefixed('proposer', nfcBytes(req.proposer))
  buf.writeInt64(req.createdAt)
  buf.writeInt64(req.expiry)

  if (req.metaHash.length !== FIXED_HASH_LEN)
    throw new InvalidEnvelopeError(`metaHash is ${req.metaHash.length} bytes, want ${FIXED_HASH_LEN}`)
  buf.write(req.metaHash)

  return buf.bytes()
}

/** SHA-256(canonicalBytes(req)) — the 32-byte value the proposer signs (§2.4). */
export function envelopeDigest(req: SigningRequest): Uint8Array {
  return sha256(canonicalBytes(req))
}

File: internal/contract/canonical.go
package contract

import (
	"bytes"
	"crypto/sha256"
	"encoding/binary"
	"encoding/hex"
	"fmt"
	"math"

	"golang.org/x/text/unicode/norm"
)

// canonicalDomain is the fixed domain-separation prefix of the envelope
// preimage (docs/spec/envelope-canonical.md §2.2):
//
//	DOMAIN = ASCII("TSS-ENVELOPE-CANONICAL-v1") ‖ 0x00
//
// The trailing 0x00 terminates the constant; the "v1" tag versions this
// canonicalization scheme and is orthogonal to SigningRequest.Version
// (docs/design/contract/protocol.md:83-87).
var canonicalDomain = append([]byte("TSS-ENVELOPE-CANONICAL-v1"), 0x00)

// fixedHashLen is the mandatory byte length of digest32 and metaHash
// (docs/spec/envelope-canonical.md §2.3): SHA-256 output, structurally fixed,
// length-prefix-free but strictly length-validated.
const fixedHashLen = 32

// maxLenPrefix is the largest value a uint32 big-endian length prefix can hold
// (docs/spec/envelope-canonical.md §2.3 rule 2).
const maxLenPrefix = math.MaxUint32

// appendUint32 appends v as 4-byte big-endian, the shared length-prefix
// encoding used by the envelope and senderAuth preimages.
func appendUint32(b []byte, v uint32) []byte {
	return binary.BigEndian.AppendUint32(b, v)
}

// CanonicalBytes builds the canonical preimage P of an envelope
// (docs/spec/envelope-canonical.md §2.1):
//
//	P = DOMAIN ‖ F(version) ‖ F(requestId) ‖ F(groupId) ‖ F(chain)
//	      ‖ F(unsignedTx) ‖ F(digest32) ‖ F(proposer)
//	      ‖ F(createdAt) ‖ F(expiry) ‖ F(metaHash)
//
// businessInfo never enters P; its integrity is carried by metaHash (which is
// in P). proposerSig never enters P; it is the signature over P. The bytes are
// derived purely from logical field values, never from JSON or protobuf wire
// bytes, so JSON-submitted and protobuf-delivered copies of the same logical
// envelope yield identical preimages (S-001 §8). Any field that violates the
// fixed-length / range contract yields ErrInvalidEnvelope, which the verifier
// treats as a rejection (docs/design/contract/protocol.md:25).
func CanonicalBytes(req *SigningRequest) ([]byte, error) {
	if req == nil {
		return nil, fmt.Errorf("%w: nil request", ErrInvalidEnvelope)
	}
	var buf bytes.Buffer
	buf.Write(canonicalDomain)

	writeUint64(&buf, req.Version)

	rid, err := uuidBytes(req.RequestID)
	if err != nil {
		return nil, err
	}
	buf.Write(rid)

	if err := writeLenPrefixed(&buf, "groupId", nfcBytes(req.GroupID)); err != nil {
		return nil, err
	}
	if err := writeLenPrefixed(&buf, "chain", nfcBytes(req.Chain)); err != nil {
		return nil, err
	}
	if err := writeLenPrefixed(&buf, "unsignedTx", req.UnsignedTx); err != nil {
		return nil, err
	}

	if len(req.Digest32) != fixedHashLen {
		return nil, fmt.Errorf("%w: digest32 is %d bytes, want %d", ErrInvalidEnvelope, len(req.Digest32), fixedHashLen)
	}
	buf.Write(req.Digest32)

	if err := writeLenPrefixed(&buf, "proposer", nfcBytes(req.Proposer)); err != nil {
		return nil, err
	}

	writeInt64(&buf, req.CreatedAt)
	writeInt64(&buf, req.Expiry)

	if len(req.MetaHash) != fixedHashLen {
		return nil, fmt.Errorf("%w: metaHash is %d bytes, want %d", ErrInvalidEnvelope, len(req.MetaHash), fixedHashLen)
	}
	buf.Write(req.MetaHash)

	return buf.Bytes(), nil
}

// EnvelopeDigest returns SHA-256(CanonicalBytes(req)) — the 32-byte value the
// proposer signs and verifiers re-derive (docs/spec/envelope-canonical.md
// §2.4).
func EnvelopeDigest(req *SigningRequest) ([32]byte, error) {
	p, err := CanonicalBytes(req)
	if err != nil {
		return [32]byte{}, err
	}
	return sha256.Sum256(p), nil
}

// writeUint64 appends an integer as fixed 8-byte big-endian
// (docs/spec/envelope-canonical.md §2.3: integers are fixed-width big-endian).
func writeUint64(buf *bytes.Buffer, v uint64) {
	var b [8]byte
	binary.BigEndian.PutUint64(b[:], v)
	buf.Write(b[:])
}

// writeInt64 appends a signed integer as fixed 8-byte big-endian (two's
// complement). createdAt/expiry are unix milliseconds
// (docs/spec/envelope-canonical.md §2.3 rule 1).
func writeInt64(buf *bytes.Buffer, v int64) {
	writeUint64(buf, uint64(v))
}

// writeLenPrefixed appends a uint32 big-endian length prefix followed by the
// raw bytes (docs/spec/envelope-canonical.md §2.3 rule 2/3). The prefix
// prevents concatenation ambiguity between adjacent variable-length fields.
func writeLenPrefixed(buf *bytes.Buffer, field string, b []byte) error {
	if uint64(len(b)) > maxLenPrefix {
		return fmt.Errorf("%w: %s exceeds uint32 length", ErrInvalidEnvelope, field)
	}
	var lp [4]byte
	binary.BigEndian.PutUint32(lp[:], uint32(len(b)))
	buf.Write(lp[:])
	buf.Write(b)
	return nil
}

// nfcBytes returns the UTF-8 bytes of s after Unicode NFC normalization
// (docs/spec/envelope-canonical.md §2.3 rule 2), so visually identical strings
// produced by different input methods canonicalize identically across parties.
func nfcBytes(s string) []byte {
	return norm.NFC.Bytes([]byte(s))
}

// uuidBytes parses an RFC 4122 UUID string into its 16 raw bytes
// (docs/spec/envelope-canonical.md §2.3: requestId is the binary 16 bytes,
// hex case-insensitive). It accepts the hyphenated 8-4-4-4-12 form and the
// 32-hex unhyphenated form; anything else is an invalid envelope.
func uuidBytes(s string) ([]byte, error) {
	var hexStr string
	switch len(s) {
	case 36:
		if s[8] != '-' || s[13] != '-' || s[18] != '-' || s[23] != '-' {
			return nil, fmt.Errorf("%w: malformed requestId UUID", ErrInvalidEnvelope)
		}
		hexStr = s[0:8] + s[9:13] + s[14:18] + s[19:23] + s[24:]
	case 32:
		hexStr = s
	default:
		return nil, fmt.Errorf("%w: requestId is not a UUID (len %d)", ErrInvalidEnvelope, len(s))
	}
	out := make([]byte, 16)
	if _, err := hex.Decode(out, []byte(hexStr)); err != nil {
		return nil, fmt.Errorf("%w: requestId not hex: %w", ErrInvalidEnvelope, err)
	}
	return out, nil
}
